File: software-cracking-10weeks/code/week04/jump_sim.cpp
// 04_jump_sim.cpp — Mô phỏng CMP + cờ ZF + nhảy JE/JNE
// Mục tiêu: hiểu vì sao JE/JNE nhảy hay không dựa vào cờ ZF.
// An toàn: thuần tính toán.  g++ -std=c++17 -O2 jump_sim.cpp -o demo
#include <iostream>
#include <cstdint>

// Mô phỏng một CPU tối giản với thanh ghi + cờ.
struct Cpu { uint32_t eax = 0, ebx = 0; bool zf = false; };

void cmp(Cpu& c, uint32_t a, uint32_t b) { c.zf = (a == b); }   // CMP a,b -> ZF=1 nếu bằng
bool je(Cpu& c)  { return c.zf; }                                // JE nhảy khi ZF==1
bool jne(Cpu& c) { return !c.zf; }                               // JNE nhảy khi ZF==0

int main() {
    const uint32_t serial = 0x1337;
    for (uint32_t input : {0x1337u, 0x1234u}) {
        Cpu c; cmp(c, input, serial);
        std::cout << "input=0x" << std::hex << input
                  << "  correct=0x" << serial << std::dec
                  << "  -> ZF=" << c.zf << "\n";
        std::cout << "   JE  (jump if equal)     : " << (je(c) ? "NHẢY (nhánh đúng)" : "không nhảy") << "\n";
        std::cout << "   JNE (jump if not equal) : " << (jne(c) ? "NHẢY (nhánh sai)" : "không nhảy") << "\n";
        std::cout << std::string(50, '-') << "\n";
    }
    std::cout << "Kết luận: 'đảo nhảy' = đổi JE<->JNE hoặc lật cờ ZF (1<->0).\n";
    return 0;
}

File: software-cracking-10weeks/code/week04/jump_sim.py
# -*- coding: utf-8 -*-
"""
04_jump_sim.py — Mô phỏng CMP + cờ ZF + lệnh nhảy có/không điều kiện.

Mục tiêu: giúp hiểu vì sao JE/JNE nhảy hay không nhảy dựa vào cờ ZF.
Đầu vào : không (chạy demo có sẵn).
Đầu ra : bảng trạng thái thanh ghi/cờ và quyết định nhảy.
An toàn : chỉ mô phỏng bằng dữ liệu, không đụng CPU thật.
"""


def cpu_step(reg_a: int, reg_b: int) -> dict:
    """Mô phỏng lệnh CMP a, b: đặt ZF=1 nếu a == b."""
    zf = 1 if reg_a == reg_b else 0
    return {"ZF": zf}


def je_taken(zf: int) -> bool:
    """JE (jump if equal) nhảy khi ZF == 1."""
    return zf == 1


def jne_taken(zf: int) -> bool:
    """JNE (jump if not equal) nhảy khi ZF == 0."""
    return zf == 0


def main() -> None:
    print("Mô phỏng kiểm tra serial:  CMP input, correct_serial")
    print("=" * 55)

    correct_serial = 0x1337
    for user_input in (0x1337, 0x1234):
        # CMP user_input, correct_serial
        flags = cpu_step(user_input, correct_serial)
        print(f"input=0x{user_input:04X}  correct=0x{correct_serial:04X}  -> ZF={flags['ZF']}")
        print(f"   JE  (jump if equal)      : {'NHẢY (nhánh đúng)' if je_taken(flags['ZF']) else 'không nhảy'}")
        print(f"   JNE (jump if not equal)  : {'NHẢY (nhánh sai)' if jne_taken(flags['ZF']) else 'không nhảy'}")
        print("-" * 55)

    print("Kết luận: để 'đảo nhảy', ta đổi JE<->JNE hoặc lật cờ ZF (1<->0).")


if __name__ == "__main__":
    main()
